### PR TITLE
[Draft] (Partial) Model performance evaluation script!

### PR DESCRIPTION
**Pull request by bwhitesell**
## What does this PR do?

This PR drafts a model performance evaluation script that is consistent with the implementation described in the first portion of section 6.3 of the alphafold3 supplemental information document.

### Details 🔬

At this stage, i've avoided as many packaging choices as possible, so the CLI control atm just leverages argparse and the multiprocessing/hardware acceleration is just done through the wrapped native torch abstractions and uses a single device for model inference - probably will .

While the main inference loop benefits from the hardware management mentioned above -- evaluation is performed via loop in a serial fashion by a single python process. Since inference time is likely to be the bottleneck by one to several orders of magnitude, I think it’s unlikely the code will need to be modified for performance reasons etc when being run on full sets of evaluation data.

Inference is performed in batch to benefit again from parallelism, so the first stage of the evaluation process is unbatching and unpadding the tensors that compose the sample to evaluate as IMO - the unpadded versions are slightly easier to reason about.

The BatchedAtomInput object we’re fed from dataloading exposes mapping tensors at the token level -- but to perform the swaps of atom positions described in the paper-- we need the mapping at the atomic level. We use the atom lengths per token idx to reconstruct all needed token mappings as atom-lvl mappings.

“Identical” asym ids are then grouped via some logic distinct to the evaluation script -- the entity type of each token in the asym id is decomposed and counted, the total counts of tokens per entity within the asym id are summed and the kv pairing of the entity ids and their counts are used to uniquely identify identical asym_ids. This is done because the tokens that compose an asym id can be from multiple entity ids -- in the case for example where a capping ion is generally treated as part of a chain.

For each of the identical identical asym id groups -- we perform an exhaustive or greedy search consistent with the implementation described in the paper. The greedy search uses simulated annealing, but the adaptation of the algorithm to the case here is not described in the paper. Some liberal assumptions were made surrounding what it meant to sample “neighboring” regions and so on. I also have made, what I _think_ are smart hyperparameter choices based on some simple hypothetical scenarios, but these may need to be tuned once trained model weights are provided.

I mainly just wanted to check in with the above and get any feedback or suggested changes before I plowed ahead. Thanks! ✌️❤️

### CODE REVIEW

**Comment by amorehead** [comment]
Hey, @bwhitesell. Stellar work so far on this PR! Briefly scanning through it, it looks like it's on the right track to the best of my understanding. Thanks again for spearheading this important effort!

**Comment by bwhitesell** [comment]
> Hey, @bwhitesell. Stellar work so far on this PR! Briefly scanning through it, it looks like it's on the right track to the best of my understanding. Thanks again for spearheading this important effort!

Sweeeet. I'll keep at it. Based on our convo -- i'm gonna skip over the ligand symmetries assignment for the moment, move onto the metric calculations. Think I need to also integrate the 5 sampled predictions per complex logic in here too. Appreciate the look over! 🚀

**Comment by amorehead** [comment]
Hi, @bwhitesell. I am about to start digging into the code you've pushed so far to see what will be the easiest way to integrate it with the existing sampling logic we've already got built (e.g., https://github.com/amorehead/alphafold3-pytorch-lightning-hydra/blob/98c142bef3e70ee2f6620970df02118c66d6d7ea/alphafold3_pytorch/models/alphafold3_module.py#L230). Just a heads up.

**Comment by bwhitesell** [comment]
> Hi, @bwhitesell. I am about to start digging into the code you've pushed so far to see what will be the easiest way to integrate it with the existing sampling logic we've already got built (e.g.,
> 
> https://github.com/amorehead/alphafold3-pytorch-lightning-hydra/blob/98c142bef3e70ee2f6620970df02118c66d6d7ea/alphafold3_pytorch/models/alphafold3_module.py#L230
> 
> ). Just a heads up.

Sounds awesome @amorehead!  Had a deadline I had to meet at work this last week so didn't get a chance to move this along much -- just started a lil work to identify interface atoms from the tensors reprs instead of the mmcif abstractions. I'm plannin on getting some in tonight tho.

I stayed away from the sampling thusfar cuz it was implemented in the lightning validation step. If we want the eval to stay a standalone script -- might be worth abstracting out the logic into a common util or somethin. 

One note to keep in mind -- each evaluation metric category defined in section 6.3 has its own logic for choosing the "best" seed.

**Comment by bwhitesell** [comment]
Hey 👋 @amorehead -- makin some more progress. Wanted to sync up!

Questions:

- I think it'd be good to move into implementing the sampling so we can start building the score aggregations. Any progress u made that I should be aware of or thoughts on approach? Might need to restructure things a little. Also -- its a little unclear what the exact mapping of metrics to interfaces/chains to me.

- When describing at the bottom of 6.3, the sample selection logic suggests using "highest interface confidence" and "highest entity confidence", any thoughts on a calculation for those? Sample that minimizes mean plddt over tokens in the interface / entity?

- We shootin for implementing this ligand resolution logic? I think we need to inorder to get good eval metrics for entity metrics on ligands and interface metrics on NA/PROT - Ligand interfaces. If so -- I think we'd need the rdkit objs and a mapping from them to the tokens they create (tokens since these r ligands), we could re-process the input file but im thinking maybe it's easier to jump into the `AtomInput`/`BatchedAtomInput` objects as theyre adapted from the `MoleculeInput` and just write the mapping as we create the tokens from the atom counts?

- DockQ, I can't find an "easy" out of the box impl in python that would nicely fit into the tensor representations we're using in the rest of the file. The seemingly authoritative [implementation](https://github.com/bjornwallner/DockQ) expects pdb/mmcif input both for the ground truth and predicted values -- we'd need to like copy the input file, inject the predicted coordinates into it, eval etc. There also I think will be minor discrepancies in the interface definitions that we'd need to put up with. Seems a little hairy, thinking about skipping for the moment.

**Comment by amorehead** [comment]
Hi, @bwhitesell. Thank you very much for diving deep into this evaluation script for us.

To be on the safe side, I'd recommend we put this implementation on hold for the time being. After reviewing the AF3 paper more closely, I think there are many other (e.g., baseline method comparison) factors that will complicate model evaluation, necessitating approaching evaluation in a different way. I'm thinking I might try putting together a relatively simple evaluation script on my own and can share my ideas once they are more concretized.

Once again, thank you for taking the initiative on this last task. I think your ligand conformation matching logic will be useful in the final evaluation script. Now I just need to finalize what it should look like and how it should be designed.

Best,
Alex

**Comment by bwhitesell** [comment]
> Hi, @bwhitesell. Thank you very much for diving deep into this evaluation script for us.
> 
> To be on the safe side, I'd recommend we put this implementation on hold for the time being. After reviewing the AF3 paper more closely, I think there are many other (e.g., baseline method comparison) factors that will complicate model evaluation, necessitating approaching evaluation in a different way. I'm thinking I might try putting together a relatively simple evaluation script on my own and can share my ideas once they are more concretized.
> 
> Once again, thank you for taking the initiative on this last task. I think your ligand conformation matching logic will be useful in the final evaluation script. Now I just need to finalize what it should look like and how it should be designed.
> 
> Best, Alex

Damn ❤️‍🩹okay. Thanks @amorehead  for taking a look over and please lmk if there’s anything else I can do, im rly eager to learn and find stuff I can help on. If you noticed something that seems dumb or didnt make sense pls lmk. Fwiw, concurrency pattern was gonna be, all processes r dedicated to dataloading except the main/parent process that uses a threadpool (1 worker per gpu) workers submit batches for inference, wait for the results, runs eval on the predictions etc.
